### PR TITLE
added sriov probe

### PR DESCRIPTION
**Pull request by sergeyglazyrindev**
responsibilites are:
1. take care of realtime changes in vf configuration
2. take care of realtime change for physical devices which provide sriov function
3. add all vfs currently configured on system
4. update state of the physical interface which provide sriov function
Partial-Bug: #SOFTID-200

Change-Id: I7736e052460ec83489c7360722d22e345f2d1960
Signed-off-by: Sergey Glazyrin <s.glazyrin@partner.samsung.com>

### CODE REVIEW

**Comment by sergeyglazyrindev** [comment]
Later on I'll add tests.

**Comment by sergeyglazyrindev** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by sergeyglazyrindev** [comment]
run all

File: common/observers/fs/fs.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// set of constants which almost mimics fsnotify type of events, except ChangedContent
+const (
+	Create Op = 1 << iota
+	Write
+	Remove
+	Rename
+	Chmod
+	ChangedContent
```
Comment (safchain) [inline common/observers/fs/fs.go]: only this one is used why ?

File: common/observers/fs/fs.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+func (w *Watcher) checkFileContentEvents() {
+	for {
+		if w.isRunning() {
+			return
+		}
+		w.Lock()
+		for filePath, fileInfo := range w.fileInfo {
+			data, err := ioutil.ReadFile(filePath)
```
Comment (safchain) [inline common/observers/fs/fs.go]: It seems to be inefficient to read the whole files each 5 seconds to check if something changed. I think that it is because you are reading /sys and inotify doesn't work. So I think we should rename or move this to specific place to avoid using this for regular File systems.

File: common/observers/fs/fs.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		}
+		w.Unlock()
+		time.Sleep(5 * time.Second)
+	}
+}
+
+// Start starts the observer
+func (w *Watcher) Start() {
+	if atomic.CompareAndSwapInt64(&w.state, common.StoppedState, common.RunningState) {
+		defer w.checkFileContentEvents()
```
Comment (safchain) [inline common/observers/fs/fs.go]: why defer here ?

File: topology/probes/netlink/handlers/abstraction.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+type LinkHandler interface {
+	ProceedNewLink(link netlink.Link)
+	Start()
+	Stop()
+}
+
+// AbstractLinkHandler defines structure for general type of handlers. The purpose is to simplify netlink.go
+// which is already so huge
+type AbstractLinkHandler struct {
+	LinkHandler
```
Comment (safchain) [inline topology/probes/netlink/handlers/abstraction.go]: not sure this line is needed here

File: common/observers/fs/fs.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+			if !ok {
+				continue
+			}
+			eventHandler(Event{
+				Name: filePath,
+				Op:   fsnotify.Op(ChangedContent),
+			})
+		}
+		w.Unlock()
+		time.Sleep(5 * time.Second)
```
Comment (safchain) [inline common/observers/fs/fs.go]: could be nice to make the duration configurable

File: topology/probes/libvirt/libvirt.go [hunk 6]
```diff
@@ -28,13 +28,17 @@ import (
 	"context"
 	"encoding/xml"
 	"fmt"
+	"strconv"
 	"sync"
 
-	"github.com/skydive-project/skydive/config"
+	"github.com/vishvananda/netlink"
```
Comment (safchain) [inline topology/probes/libvirt/libvirt.go]: can you send another PR for libvirt modifications

File: topology/probes/netlink/handlers/abstraction.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package handlers
```
Comment (safchain) [inline topology/probes/netlink/handlers/abstraction.go]: copyright missing

File: topology/probes/netlink/handlers/sriov.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
+package handlers
```
Comment (safchain) [inline topology/probes/netlink/handlers/sriov.go]: copyright missing

File: topology/probes/netlink/handlers/sriov.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// ProceedNewLink starts tracking changes for specific link
+func (h *SriovLinkHandler) ProceedNewLink(link netlink.Link) {
+	h.watcher.AddNewPath(fmt.Sprintf("/sys/class/net/%s/operstate", link.Attrs().Name))
+}
+
+// OnChangedFile being triggered when file operstate changed in /sys/ linux file structure
+func (h *SriovLinkHandler) OnChangedFile(details fs.Event) {
+	re := regexp.MustCompile("/sys/class/net/(.+)/operstate")
```
Comment (safchain) [inline topology/probes/netlink/handlers/sriov.go]: move this as a const to avoid recompiling it each time

File: common/observers/fs/fs.go (hunk 3 above)
Comment (safchain) [reply common/observers/fs/fs.go]: should we start a goroutine ?

File: topology/probes/netlink/netlink.go [hunk 10]
```diff
@@ -79,6 +80,7 @@ type NetNsProbe struct {
 	state                int64
 	wg                   sync.WaitGroup
 	quit                 chan bool
+	subprobes            map[string]*handlers.AbstractLinkHandler
```
Comment (safchain) [inline topology/probes/netlink/netlink.go]: should be map[string]LinkHandler so the interface not the implementation

File: topology/probes/sriov/no_sriov.go [hunk 11]
```diff
@@ -0,0 +1,25 @@
+// +build !linux
```
Comment (safchain) [inline topology/probes/sriov/no_sriov.go]: header missing

File: topology/probes/sriov/sriov.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+	Graph  *graph.Graph
+	Root   *graph.Node
+	Cancel context.CancelFunc // cancel function
+}
+
+// MANAGER is a const to be used as a string to describe who handles sriov nodes
+const MANAGER = "vfio"
+
+// GenerateNameForNodeFromPhysInterfaceNameAndVFID - generates a name for vf based on physical device name and id of of this vf
+func GenerateNameForNodeFromPhysInterfaceNameAndVFID(PhysName string, ID string) string {
```
Comment (safchain) [inline topology/probes/sriov/sriov.go]: keep local variable with lowercase first letter

File: topology/probes/sriov/sriov.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+	filter := filters.NewAndFilter(
+		filters.NewTermStringFilter("Manager", MANAGER),
+	)
+	m := graph.NewElementFilter(filter)
+	return h.Graph.GetNodes(m)
+}
+
+func (h *Probe) periodicChecker(ctx context.Context) {
+	for ctx.Err() == nil {
+		linkList, err := netlink.LinkList()
```
Comment (safchain) [inline topology/probes/sriov/sriov.go]: I would prefer to have it event based not using polling pattern. Can ProceedNewLink not be used ? The handler could be extended to handled deletion too

File: common/observers/fs/fs.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+// WatcherCallback described the interface for callback on matched event
+type WatcherCallback = func(details Event)
+
+// Op mimics fsnotify.Op for backward compatibility with fsnotify
+type Op fsnotify.Op
+
+// Event mimics fsnotify.Event for backward compatibility with fsnotify
+type Event fsnotify.Event
```
Comment (safchain) [inline common/observers/fs/fs.go]: I do not see where you are using fsnotify and why you have to ? I would define my own event structure to pass the content in the event. It would avoid the read it twice, in the checkFileContentEvents and in the srio-v callback

File: common/observers/fs/fs.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+	logging.GetLogger().Debugf("Register event handler: %d", eventType)
+	w.eventHandlers[eventType] = cb
+}
+
+// AddNewPath starts tracking changes for new path
+func (w *Watcher) AddNewPath(path string) {
+	logging.GetLogger().Debugf("track changes for the file: %s", path)
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		logging.GetLogger().Errorf("Error while reading content of the file: %s, Err: %s", path, err)
```
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: done

File: common/observers/fs/fs.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
+package fs
```
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: I don't want to make from common package a GOD package. That's why I created extra subfolder. Don't you think it's better to have small packages (ie namespaces) than huge namespaces ?

File: common/observers/fs/fs.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+		Path:    path,
+		Content: data,
+	}
+	if err != nil {
+		logging.GetLogger().Errorf("Error while adding path to fs watcher: %s. Err: %s", path, err)
+	}
+}
+
+func (w *Watcher) checkFileContentEvents() {
+	for {
```
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: fixed.

File: common/observers/fs/fs.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+		if w.isRunning() {
```
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: fixed.

File: common/observers/fs/fs.go (hunk 1 above)
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: basically, my problem I tried to solve is following: when operator executes in bash console ifconfig SRIOV_PHYS_DEVICE up/down then kernel event doesn't occur, so we are not able to react appropriately. First of all I tried to use fsnotifier to watch for the file: /sys/class/net/SRIOV_PHYS_DEVICE/operstate but it doesn't trigger any fsnotify events, so I added my own event: ChangedContent which should be used very carefully and only when you really need it. But this module simplifies usage of fsnotify... Later on when we will need it it's very easy to add a separated handler for fsnotify events and using interface of this module it's very easy to subscribe to different events. I added all these constants for backward compatibility with fsnotify package.

File: common/observers/fs/fs.go [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+// OnEvent registers handler for specific type of event, simplifies life for developer
+func (w *Watcher) OnEvent(eventType Op, cb WatcherCallback) {
+	w.Lock()
+	defer w.Unlock()
+	logging.GetLogger().Debugf("Register event handler: %d", eventType)
+	w.eventHandlers[eventType] = cb
+}
+
+// AddNewPath starts tracking changes for new path
+func (w *Watcher) AddNewPath(path string) error {
```
Comment (lebauce) [inline common/observers/fs/fs.go]: nit: s/AddNewPath/AddPath/

File: common/observers/fs/fs.go [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+	w.Lock()
+	defer w.Unlock()
+	_, doWeTrackIt := w.fileInfo[path]
```
Comment (lebauce) [inline common/observers/fs/fs.go]: nit:
```
if _, tracked := w.fileInfo[path]; tracked {
    return nil
}
```

File: topology/probes/sriov/sriov.go [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+// Probe is a Sriov probe designed to handle all vfs on server
+type Probe struct {
+	Graph   *graph.Graph
+	Root    *graph.Node
+	Cancel  context.CancelFunc // cancel function
+	watcher *fs.Watcher
+}
+
+// MANAGER is a const to be used as a string to describe who handles sriov nodes
+const MANAGER = "vfio"
```
Comment (lebauce) [inline topology/probes/sriov/sriov.go]: By convention, constants should not be uppercase in Go unless it's an acronym

File: topology/probes/sriov/sriov.go [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+	"github.com/skydive-project/skydive/common/observers/fs"
+	"github.com/skydive-project/skydive/filters"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+)
+
+// Probe is a Sriov probe designed to handle all vfs on server
+type Probe struct {
+	Graph   *graph.Graph
```
Comment (lebauce) [inline topology/probes/sriov/sriov.go]: nit: no need to export `Graph`, `Root` and `Cancel`

File: topology/probes/sriov/sriov.go [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+	Root    *graph.Node
+	Cancel  context.CancelFunc // cancel function
+	watcher *fs.Watcher
+}
+
+// MANAGER is a const to be used as a string to describe who handles sriov nodes
+const MANAGER = "vfio"
+
+// GenerateNameForNodeFromPhysInterfaceNameAndVFID - generates a name for vf based on physical device name and id of of this vf
```
Comment (lebauce) [inline topology/probes/sriov/sriov.go]: nit: this seems to be used only once (and the name is a bit long :-) ) so I suggest we just inline it in the function where it is used

File: topology/probes/sriov/sriov.go [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+				}
+				logging.GetLogger().Debugf("try to process sriov nodes and vf : %s. MAC: %s, %b", attrs.Name, vf.Mac, vf.Mac)
+				h.addVfNode(name, parentNode, &vf)
+				addedNodes[name] = true
+			}
+			if len(attrs.Vfs) > 0 {
+				h.watcher.AddNewPath(fmt.Sprintf("/sys/class/net/%s/operstate", attrs.Name))
+			}
+		}
+		allNodes := h.getAllSriovNodes()
```
Comment (lebauce) [inline topology/probes/sriov/sriov.go]: nit: 
```
for _, node := range {
    ...
}
`

File: topology/probes/sriov/sriov.go [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+		for _, node := range allNodes {
+			nodeName, _ := node.GetFieldString("Name")
+			_, ok := addedNodes[nodeName]
```
Comment (lebauce) [inline topology/probes/sriov/sriov.go]: nit:
```
if _, ok := addNodes; ok {
    continue
}
```

File: topology/probes/sriov/sriov.go (hunk 23 above)
Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: Later on it will be used in libvirt probe to connect sriov vf with libvirt vm. I'll make pull request later.

File: topology/probes/sriov/sriov.go [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+
+func (h *Probe) getAllSriovNodes() []*graph.Node {
+	filter := filters.NewAndFilter(
+		filters.NewTermStringFilter("Manager", MANAGER),
+	)
+	m := graph.NewElementFilter(filter)
+	return h.Graph.GetNodes(m)
+}
+
+func (h *Probe) periodicChecker(ctx context.Context) {
```
[the remark this answers is not in the document]
Comment (safchain) [reply topology/probes/sriov/sriov.go]: Yes a ticker and a select to watch ctx.Done()

File: topology/probes/sriov/sriov.go (hunk 23 above)
Comment (lebauce) [reply topology/probes/sriov/sriov.go]: ok

File: topology/probes/sriov/sriov.go [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+		var addedNodes = make(map[string]bool)
+		for _, link := range linkList {
+			attrs := link.Attrs()
+			parentNode := h.graph.LookupFirstChild(h.root, graph.Metadata{"Name": attrs.Name, "IfIndex": int64(attrs.Index)})
+			if parentNode == nil {
+				continue
+			}
+			for _, vf := range attrs.Vfs {
+				name := GenerateNameForNodeFromPhysInterfaceNameAndVFID(attrs.Name, strconv.Itoa(vf.ID))
+				if allZero(vf.Mac) {
```
Comment (lebauce) [inline topology/probes/sriov/sriov.go]: the call the allZero should probably be done before `GenerateNameForNodeFromPhysInterfaceNameAndVFID`

Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: done

File: topology/probes/sriov/sriov.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: pushed changes, now testing it on my environment

File: topology/probes/sriov/sriov.go (hunk 25 above)
Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: done

File: topology/probes/sriov/sriov.go (hunk 24 above)
Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: done

File: topology/probes/sriov/sriov.go (hunk 22 above)
Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: done

File: topology/probes/sriov/sriov.go (hunk 21 above)
Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: donme

File: common/observers/fs/fs.go (hunk 20 above)
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: done

File: common/observers/fs/fs.go (hunk 19 above)
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: done

File: common/observers/fs/fs.go [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+			fileInfo.Content = data
+			w.fileInfo[filePath] = fileInfo
+			eventHandler, ok := w.eventHandlers[ChangedContent]
+			if !ok {
+				continue
+			}
+			eventHandler(Event{
+				Event: fsnotify.Event{
+					Name: filePath,
+					Op:   fsnotify.Op(ChangedContent),
```
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: done

File: common/observers/fs/fs.go (hunk 1 above)
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: done

File: common/observers/fs/fs.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply common/observers/fs/fs.go]: done

File: agent/probes.go [hunk 29]
```diff
@@ -114,6 +115,12 @@ func NewTopologyProbeBundleFromConfig(g *graph.Graph, hostNode *graph.Node) (*pr
 				return nil, fmt.Errorf("Failed to initialize Libvirt probe: %s", err)
 			}
 			probes[t] = libvirt
+		case "sriov":
```
Comment (safchain) [inline agent/probes.go]: can you activate it in the tests even if we don't have srvi-o on CI it will test that it starts properly
https://github.com/skydive-project/skydive/blob/master/tests/tests.go#L91

Comment (sergeyglazyrindev) [reply agent/probes.go]: added

File: topology/probes/sriov/sriov.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (sergeyglazyrindev) [reply topology/probes/sriov/sriov.go]: tested. seems like it works.

**Comment by sergeyglazyrindev** [comment]
hey @safchain 
Btw, please don't merge it yet, I'll contribute it from account assigned to samsung.com email. I didn't know the exact procedure, that's why I made this pull request from my private account, but it seems like the proper way to contribute according to my company regulations is to contribute from private github account but assigned to my samsung.com email

**Comment by safchain** [comment]
@sergeyglazyrindev @lebauce ok, tell us when ready, thanks for addressing the comments :)
